File: Object/C++/animals/animal.cpp
#include <string>
#include <optional>
#include "animal.h"
#include <stdio.h>

Animal::Animal(int anID, std::string anName): id(anID), name(anName) {
    owner = std::nullopt;
}

Animal::Animal(int anID, std::string anName, std::string anOwner): id(anID), name(anName), owner(anOwner) {}

int Animal::ID() const {
    return id;
}

std::string Animal::GetName() const {
    return name;
}

std::string Animal::GetOwner() const {
    return owner.value_or("No owner");
}

void Animal::SetName(std::string newName) {
    name = newName;
}

void Animal::Identify() const {
    printf("My name is %s.\n", name.c_str());
    if (owner == std::nullopt) {
        printf("I have no owner.\nMy id is %d.\n", id);
    } else {
        printf("My owner is %s.\nMy id is %d.\n", owner.value().c_str(), id);
    }
}

File: Object/Java/animals/Animal.java
import java.util.Optional;

abstract class Animal {
    private int id;
    private String name;
    private Optional<String> owner;

    public Animal(int anID, String anName) {
        id = anID;
        name = anName;
        owner = Optional.empty();
    }

    public Animal(int anID, String anName, String ownName) {
        id = anID;
        name = anName;
        owner = Optional.ofNullable(ownName);
    }

    public int ID(){
        return id;
    }
    public String GetName(){
        return name;
    }
    public String GetOwner(){
        return owner.orElse("No owner");
    }
    public void SetName(String newName){
        name = newName;
    }
    public void SetOwner(String newOwner){
        owner = Optional.ofNullable(newOwner);
    }
    public void RemoveOwner() {
        owner = Optional.empty();
    }

    abstract void MakeNoise();
    public void Identify(){
        System.out.println("My name is " + name + ".");
        if (owner.isEmpty()) {
            System.out.println("I have no owner.\nMy id is " + id +".\n");
        } else {
            System.out.println("My owner is " + owner.get() +".\nMy id is " + id +".\n");
        }
    }

}

